### ISSUE TITLE
get_users()

### ISSUE THREAD

**Issue by yeled** [napalm-automation-community/napalm-opengear#1]
migrating this issue over from https://github.com/yeled/napalm-opengear/issues/16#issuecomment-544361030

> All passwords are MD5 hashed in /etc/shadow and not stored in config.xml, _except_ for users that are in the PSTN _dialin_ group (we recommend using a separate auxiliary account for dialin auth).
> 
> You can set the **config.users.userX.password** field to whatever you want writing directly to the shadow file's encrypted password field, or you can set **config.users.userX.plaintext_password** and the configurator will encrypt it for you. In both cases the passwords are removed from config.xml by the configurator as it runs.
> 
> The examples at the end of KB article might be useful: [opengear.zendesk.com/hc/en-us/articles/216371303-Managing-local-users-from-the-command-line](https://opengear.zendesk.com/hc/en-us/articles/216371303-Managing-local-users-from-the-command-line)
> 
> If you remove and then re-add a user and then finally run the configurator, I would've thought the user's homedir would be untouched.

First I will implement get_users() and then come up with some method for deleting them via some magic config and  `/etc/scripts/user-del`

**Comment by yeled** [comment]
@r0bertw I think I finally understand what you're saying now.

However, upon `config -a` the below should work, and `config.users.user3.plaintext_password` should disappear from `config.xml`. But it doesn't! In fact `/etc/shadow` doesn't have those users anymore.

Am I missing something that `config -r users` needs?

I have this Jinja which deletes all the `config.users` (root is not affected), then outputs this config to be set:

**consoles.yaml**
```yaml
user2: XXX
user2_password: YYYXXX
user3: YYY
user3_password: XXXYYY
```

**aaa.j2**
```jinja2
config.users

config.users.user2.username {{ user2 | env_lookup }}
config.users.user2.groups.group1 admin
config.users.user2.groups.group2 pmshell
config.users.user2.plaintext_password {{ user2_password | env_lookup }}

config.users.user3.username {{ user3 | env_lookup }}
config.users.user3.groups.group1 admin
config.users.user3.groups.group2 pmshell
config.users.user3.plaintext_password {{ user3_password | env_lookup }}
```

` | env_lookup` is a simple filter plugin that grabs the environment variable secret password.


```
config.users.user2.username XXX
config.users.user2.groups.group1 admin
config.users.user2.groups.group2 pmshell
config.users.user2.plaintext_password YYYXXX

config.users.user3.username YYY
config.users.user3.groups.group1 admin
config.users.user3.groups.group2 pmshell
config.users.user3.plaintext_password XXXYYY
```

**Comment by r0bertw** [comment]
Are you updating `config.users.total` before running the configurator?  Same goes for `config.users.userX.groups.total`.

If that's the issue, there's a brief note about [config List Operations in this KB article](https://opengear.zendesk.com/hc/en-us/articles/216371123-Configuring-from-the-command-line).

**Comment by yeled** [comment]
Thanks @r0bertw -- AFAICT, this solves all user management without diving into special `user-mod` calls.

It deletes all users then re-instates 2 users, `user2` and `user3`.

```jinja2
config.users

config.users.total 3

config.users.user1.groups.total 7
config.users.user1.username {{ user1 }}
config.users.user1.plaintext_password {{ user1_password | env_lookup }}

config.users.user2.username {{ user2 }}
config.users.user2.groups.group1 admin
config.users.user2.groups.group2 pmshell
config.users.user2.groups.group3 users
config.users.user2.groups.total 3

config.users.user2.plaintext_password {{ user2_password | env_lookup }}

{% for port in range(1, (napalm_model[4:6]|int)|default('16') + 1) %}{# port count from model name #}
config.users.user2.port{{ port }} on
{% endfor %}

config.users.user3.username {{ user3 }}
config.users.user3.groups.group1 admin
config.users.user3.groups.group2 pmshell
config.users.user3.groups.group3 users
config.users.user3.groups.total 3

config.users.user3.plaintext_password {{ user3_password | env_lookup }}

{% for port in range(1, (napalm_model[4:6]|int)|default('16') + 1) %}{# port count from model name #}
config.users.user3.port{{ port }} on
{% endfor %}
```

I am having another issue right now where napalm is erroring like this:

```
fatal: [cons-n3-1.net.sdc42-sea.github.net]: FAILED! => {"changed": false, "msg": "cannot diff config: Search pattern never detected in send_command_expect: \\<fig\\.services\\.snmp\\.roauthprotocol\\=SHA\\'\\ \\-s\\ \\'config\\.se\\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\b\\\brvices\\.snmp\\.ro"}
```

That's 25 spaces, then 25 backspaces.

I assume it's the long lines and the interaction of the opengear terminal.

**Comment by yeled** [comment]
yep, fixed in #7 

yay this works!
